pkg/internal/testhelpers/containers.go

```diff
@@ -130,118 +199,263 @@ func (s stdoutLogConsumer) Accept(l testcontainers.Log) {
+		})
+		if err != nil {
+			return nil, err
+		}
 
-	container, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
-		ContainerRequest: req,
-		Started:          false,
-	})
-	if err != nil {
-		return nil, err
-	}
+		if printLogs {
+			container.FollowOutput(stdoutLogConsumer{})
+		}
 
-	if printLogs {
-		container.FollowOutput(stdoutLogConsumer{})
-	}
+		if printLogs {
+			err = container.StartLogProducer(context.Background())
+			if err != nil {
+				fmt.Println("Error setting up logger", err)
+				os.Exit(1)
+			}
+		}
 
-	if printLogs {
-		err = container.StartLogProducer(context.Background())
+		err = container.Start(context.Background())
 		if err != nil {
-			fmt.Println("Error setting up logger", err)
-			os.Exit(1)
+			return nil, err
 		}
-	}
 
-	err = container.Start(context.Background())
-	if err != nil {
-		return nil, err
-	}
+		c.Append(func() {
+			if printLogs {
+				_ = container.StopLogProducer()
+			}
+
+			_ = container.Terminate(ctx)
+		})
+
+		// we map the access node to port 5432 and the others to other ports
+		if isDataNode {
+			// trust all incoming connections on the datanodes so we can add them to
+			// the cluster without too much hassle. We'll reload the pg_hba.conf later
+			// NOTE: the setting must be prepended to pg_hba.conf so it overrides
+			//       the latter, stricter, settings
+			code, err := container.Exec(context.Background(), []string{
+				"bash",
+				"-c",
+				"echo -e \"host all all 0.0.0.0/0 trust\n$(cat /var/lib/postgresql/data/pg_hba.conf)\"" +
+					"> /var/lib/postgresql/data/pg_hba.conf"})
+			if err != nil {
+				return nil, err
+			}
+			if code != 0 {
+				return nil, fmt.Errorf("docker exec error. exit code: %d", code)
+			}
+		}
 
-	pgHost, err = container.Host(ctx)
-	if err != nil {
-		return nil, err
+		pgHost, err = container.Host(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		pgPort, err = container.MappedPort(ctx, containerPort)
+		if err != nil {
+			return nil, err
+		}
+
+		db, err := pgx.Connect(context.Background(), PgConnectURL(defaultDB, Superuser))
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = db.Exec(context.Background(), fmt.Sprintf("CREATE USER %s WITH NOSUPERUSER CREATEROLE PASSWORD 'password'", promUser))
+		if err != nil {
+			//ignore duplicate errors
+			pgErr, ok := err.(*pgconn.PgError)
+			if !ok || pgErr.Code != "42710" {
+				return nil, err
+			}
+		}
+
+		if isDataNode {
+			// reload the pg_hba.conf settings we changed above
+			_, err = db.Exec(context.Background(), "SELECT pg_reload_conf()")
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		err = db.Close(context.Background())
+		if err != nil {
+			return nil, err
+		}
+
+		return container, nil
 	}
 
-	pgPort, err = container.MappedPort(ctx, containerPort)
-	if err != nil {
-		return nil, err
+	if multinode {
+		containerPort := nat.Port("5433/tcp")
+		_, err := startContainer(containerPort)
+		if err != nil {
+			c.Close()
```

Comment:
why not do this in a defer?

Reply:
because if we _don't_ error we still need the container around